### PR TITLE
MappingApplication improve useability

### PR DESCRIPTION
**Pull request by philbucher**
This pull request contains the following improvements:
1. Class descriptions have been added
2. The Validation of the JSON Parameters Object is now done by the MapperCommunicator. This makes more sense, especially if the MapperCommunicator is used on it's own, i.e. without a Mapper
3. The printed output has been unified, now the Echo Level is consistent across the application. This simplifies debugging a lot. An overview of the levels can be found in the class description of the MapperCommunicator
4. Fixing a compilation warning abt an uninitialized variable
5. Now also the main ModelPart can be used as input for the Mapper, the input does not necessarily have to be a SubModelPart any more

### CODE REVIEW

File: applications/MappingApplication/custom_configures/interface_object_configure.h
Comment (jcotela) [inline applications/MappingApplication/custom_configures/interface_object_configure.h]: Configuration file for *InterfaceObject*?

**Review by jcotela — APPROVED** [review]
:+1: A good clean-up of the application. By the way, there is a kratos_logger class in the core that I think would be good for the kind of very stantarized output format you are using here, but I think it is not ready for parallel output yet (all threads would print). @pooyan-dadvand can you confirm this?

**Comment by philbucher** [comment]
Corrected the Typo

I agree that using the logger would make sense, I will wait with merging until @pooyan-dadvand answers if it is ready for mpi.

**Comment by jcotela** [comment]
Well, feel free to merge, we can always add this later.
By the way, regarding your point 5, this is just what I need now for my application. How do I use it? Is it enough if you don't define the corresponding submodelpart name in JSON parameters?

**Comment by philbucher** [comment]
Ok then I merge now and add the logging from the core later. I still have some more TODOs on my list anyway.

If you want to use the main ModelPart just don't specify the "interface_submodel_part_*" parameter in the json input file. I am updating the Wiki atm.

File: applications/MappingApplication/custom_utilities/mapper.h
Comment (RiccardoRossi) [inline applications/MappingApplication/custom_utilities/mapper.h]: Hi Philip,

the behaviour of parameters is "strange" in the sense that the Parameters object holds internally a pointer to the data of the lib it is built upon. This means that the correct way of passing it is

     Parameters JsonParameters

which will behave much like as if it was a reference. 
Passing a reference is actually a memory error, so i must block your pull request for this.

The point is that the wrapping of the rapidjson library was a little tricky, as we needed to ensure that the data exists until the last Parameters object exists. This is the reason for which we hold a shared ptr to the underlying data (with the counter being updated when you pass a Parameters object)

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]

File: applications/MappingApplication/custom_utilities/mapper.h
Comment (philbucher) [reply applications/MappingApplication/custom_utilities/mapper.h]: Hi Riccardo,
Thanks for pointing this out, I didn't know.
So I should always pass the parameters by value (`Parameters JsonParameters`)? And in doing this I will create another shared pointer to the underlying data?

**Comment by RiccardoRossi** [comment]
Hi @philbucher 

yes, to understand what happens, take a look here

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/kratos_parameters.h#L130

and 

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/kratos_parameters.h#L137

**Comment by philbucher** [comment]
@RiccardoRossi is this ok now?

**Comment by pooyan-dadvand** [comment]
At this moment Logger is not MPI enabled.... I just added it to the global plan project.

**Comment by philbucher** [comment]
@jcotela could you please check my latest commit? Since @RiccardoRossi blocked this, I added another check that involved more changes than I thought.
I compute the bounding boxes of the entire ModelPart and print a warning in case they do not overlap. We had problems with th Olympic Roof, where the Models were physically in different locations. This should be avoided with this check.
Since the bounding boxes are also needed elsewhere I did some refractoring, and had to change some files
Another question abt the json Parameters, can I pass them in as `const`? It is a pointer that I do not want to modify.

File: applications/MappingApplication/custom_utilities/mapper_utilities.h
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/mapper_utilities.h]: If you are going to check both local and ghost nodes, you can simply iterate over ModelPart.Nodes()

File: applications/MappingApplication/custom_utilities/interface_object.h
Comment (jcotela) [inline applications/MappingApplication/custom_utilities/interface_object.h]: Note that Node::Cordinates already gives you the {X,Y,Z} array (as a Kratos::Point). I would recommend using that instead of creating an ad-hoc vector

**Comment by jcotela** [comment]
@philbucher I added a couple of comments. Another, more general consideration is that BoundingBox methods seem to mix `double*` and `std::vector<double>` arrays when passing around the bounding box itself. Maybe a simple container class for the BoundingBox would be a good idea to homogenize how the information is passed around? (or just deciding on a 'standard format')

File: applications/MappingApplication/custom_utilities/interface_object.h
Comment (philbucher) [reply applications/MappingApplication/custom_utilities/interface_object.h]: I know, but in other cases I don't have an underlying point. e.g. [here](https://github.com/KratosMultiphysics/Kratos/blob/feature_mappingApp_improve_useability/applications/MappingApplication/custom_utilities/mapper_utilities.h#L313), should I create a Point and pass it to the function (and change the function to accept only Points?) Or is it better to overload the function.
I would go for creating a Point, I think this shouldn't be too expensive?

File: applications/MappingApplication/custom_utilities/mapper_utilities.h
Comment (philbucher) [reply applications/MappingApplication/custom_utilities/mapper_utilities.h]: Will change it

**Comment by philbucher** [comment]
@jcotela I agree with the bounding boxes.
The problem is that for mpi I have to use c-arrays, whereas if I don't have to do this I would like to avoid it.
There is [bounding_box.h](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/spatial_containers/bounding_box.h) and [bounding_box_utilities.h](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/bounding_box_utilities.h) in the core, but those are quite outdated from what I could see. 
Bottom line, either we refractor this class (in case it is possible) or I create another Aux Class in the MappingApplication. What do you think?

File: applications/MappingApplication/custom_utilities/interface_object.h
Comment (jcotela) [reply applications/MappingApplication/custom_utilities/interface_object.h]: In the current implementation, the InterfaceObject is already a Point (it derives from it), so you can use Point::Coordinates to get an `array_1d<double,3>` containing {X,Y,Z}. I think this would be usable both by node-based and integrationpoint-based mappers. This would not require creating new containers if passed by reference, since everything is already contained in (the base class of) InterfaceObject. Do you think this could work?

**Comment by jcotela** [comment]
@philbucher On the topic of bounding boxes, I'd say we wait for things to settle down a bit with global pointers and the new search stuff. I'd propose refactoring, but it is not worth it now, given that all this might change significantly in the mid-term as the next iteration of spatial search tools is developed.

**Comment by philbucher** [comment]
@jcotela good point. As I said I hadn't planned to change so much just for a small check. How abt we merge the previous commit (48a08c4b880ca687823ddf7714516363e769aa38)?
This doesn't change the search related stuff and contains only what this pull request is created for, i.e. cleaning up
I will keep the changes locally and will get back to it when I have to refractor anyway bcs of changing search, global pointer,...

**Comment by jcotela** [comment]
@philbucher For me you can just go ahead and merge this once @RiccardoRossi considers the Parameters are passed in the right way (I'm not an expert on the details of that, so I'll leave it up to him).
The things I mentioned in the last round of comments are relatively minor, so I'd just ask you to keep them in mind for the next round of changes.

**Comment by pooyan-dadvand** [comment]
For bounding box I would definitely go for a class. Actually I'm implementing one in the [searchlab](https://github.com/pooyan-dadvand/SearchLab) which can be used as reference

**Comment by philbucher** [comment]
@pooyan-dadvand For now I will leave it, since the search with the bounding boxes will be replaces once the mpi search works. If in the future I will need it for sth else I will implement the bounding box solution

@RiccardoRossi is this correct now (the passing of the json parameters?) Or should I pass them as const?

**Comment by RiccardoRossi** [comment]
you don't need to pass them as const

File: applications/MappingApplication/custom_utilities/mapper_communicator.h
Comment (RiccardoRossi) [inline applications/MappingApplication/custom_utilities/mapper_communicator.h:404]: just a curiosity. why didn't you use the Validate method? i think that would have been more concise than this...

**Review by RiccardoRossi — APPROVED** [review]
i don't have any reason to block this. 

i am not sure of why you choosed not to use the json validate mechanism, your call anyway.

the other point i can rise is to take care not to use JsonParameters in time critical parts of the code. I am not telling you are doing it...just warning that the Parameters object is relatively slow.

File: applications/MappingApplication/custom_utilities/mapper_communicator.h
Comment (philbucher) [reply applications/MappingApplication/custom_utilities/mapper_communicator.h:404]: I validate it later
The `Has` methods I use to check the value before the validation
But I will think abt it, thanks for the comment

**Pull request merged**
